invirtualenv/plugin_base.py

```diff
@@ -162,25 +166,58 @@ def generate_wheel_packages(self, wheeldir):
+            except subprocess.CalledProcessError as error:
+                logger.warning('Exception occurred while generating wheel packages, downloading source packages')
+                if error.stdout:
+                    logger.error(error.stdout.decode())
+                if error.stderr:
+                    logger.error(error.stderr.decode())
+                cmd = self.pip_cmd + ['download', '-d', '.'] + deps
+                logger.debug('Running pip command %r to download missing packages', cmd)
+                try:
+                    output = subprocess.check_output(cmd, stderr=subprocess.STDOUT).decode()  # nosec
+                except subprocess.CalledProcessError as error:
+                    logger.warning('Exception occurred while downloading source packages')
+                    if error.stdout:
+                        logger.error(error.stdout.decode())
+                    if error.stderr:
+                        logger.error(error.stderr.decode())
+                    raise
             for filename in os.listdir('.'):
                 if filename.endswith('.whl'):
-                    if not filename.endswith('none-any.whl'):
+                    split_filename = os.path.basename(filename).split('-')
+                    file_wheel_name = filename
+                    file_wheel_version = None
+                    if len(split_filename) > 2:
+                        file_wheel_name = split_filename[0]
+                        file_wheel_version = split_filename[1]
+                    if self.noarch and not filename.endswith('none-any.whl'):
                         self.noarch = False
                     cmd = self.pip_cmd + ['hash']
                     if self.hash:
                         cmd += ['-a', self.hash]
                     cmd += [filename]
                     logger.debug('Running pip command %r to generate package hash for %r', cmd, filename)
-                    hashes[filename] = '='.join(subprocess.check_output(cmd).decode().split(os.linesep)[1].split('=')[1:])  # nosec
-                    logger.debug('Got requirements line %r', hashes[filename])
+                    hash_result = subprocess.check_output(cmd).decode()  # nosec
+                    if file_wheel_name and file_wheel_version:
+                        logger.debug(f"file_wheel_name==file_wheel_version")
+                        hashes['{file_wheel_name}=={file_wheel_version}'.format(**locals())] = '='.join(hash_result.split(os.linesep)[1].split('=')[1:])
+                        logger.debug('Got requirements line %r', hashes['{file_wheel_name}=={file_wheel_version}'.format(**locals())])
+                    else:
+                        hashes[filename] = '='.join(hash_result.split(os.linesep)[1].split('=')[1:])  # nosec
+                        logger.debug('Got requirements line %r', hashes[filename])
+        self._wheel_hashes = hashes
+        self.add_plugin_configuration()
```

Comment:
Instead of calling `self.add_plugin_configuration()` here, calling it at the beginning of `render_template_with_config`?